src/core/main.c

```diff
@@ -147,9 +149,49 @@ static NUMAPolicy arg_numa_policy;
 /* A copy of the original environment block */
 static char **saved_env = NULL;
 
+static volatile sig_atomic_t waiting_for_signal_to_reexec = 0;
+static sigjmp_buf crash_wait_reexec_jmp_buf;
+
 static int parse_configuration(const struct rlimit *saved_rlimit_nofile,
                                const struct rlimit *saved_rlimit_memlock);
 
+static void reexec_handler(int sig) {
+        waiting_for_signal_to_reexec = 1;
+}
+
+static int do_crash_wait_reexec(void) {
+        struct sigaction sa = {
+                .sa_handler = reexec_handler,
+                .sa_flags = SA_RESTART,
+        };
+        sigset_t ss;
+        int r = 0;
+
+        sigemptyset(&ss);
+        sigaddset(&ss, SIGTERM);
+        sigprocmask(SIG_UNBLOCK, &ss, NULL);
+
+        sigaction(SIGTERM, &sa, NULL);
+
+        log_emergency("Fatal crash, PID1 halted. Send SIGTERM to PID1 to trigger reexecution.");
+        waiting_for_signal_to_reexec = 0;
+        while(1) {
+                usleep(10000);
+                if (waiting_for_signal_to_reexec) {
+                        log_emergency("Jumping to manager_reexecute now...");
+                        siglongjmp(crash_wait_reexec_jmp_buf, 1);
```

Comment:
hmm, so what happens with any open fds here? they just stay open? that effectively leaks them all

Reply:
longjump() is really weird in its semantics, as we'll leak mmaps, fds and everything else here. 

I remember that I played around with execve() from the signal handler to actually reexec ourselves once upon a time, but i couldn't get it to work for pid1.

can you explain why you tried siglongjmp() here? did you play around with just reexecing pid1 properly, i.e. via execve()?